docs/design2.md

```diff
@@ -0,0 +1,109 @@
+After verifying, the gossip node signs the log root.
+
+Gossip nodes offer the following APIs:
+
+1. Get Signed Root
+
+### Log of Map Roots
+
+The log server collects signatures from the gossip nodes to produce a Signed Log Root that has a quorum of gossip signatures.
+
+The log server contains a list of map roots representing sequential snapshots of a key-value directory.
+
+Log servers offer the following APIs:
+
+1. Latest Signed Log Root
+1. Get consistency proof between log size a and b.
+1. Get log item i with inclusion proof to log size a.
+
+### Map Root Snapshots
+Each Map Root represents a snapshot of a key-value dictionary.
+The map is implemented as a sparse Merkle tree of fixed depth.
+
+* Indexes in the map are randomized and privacy protected via a Verifiable Random Function.
+* Values in the map represent the full history of values that have ever been stored at this index. This is accomplished by storing the Merkle root of a mini-log of these values.
+
+The map offers the following APIs:
+
+1. Get map value at index j with inclusion proof to snapshot a.
+
+### Value History Log
+These mini logs store not just the latest value, but also store every previous value in order.
+
+They are what users query when looking up their own key history, and they are what their peers verify in order to ensure that no history has been lost.
+
+The value log offers the following API:
+
+1. Get latest value at snapshot z with inclusion proof.
+2. Get consistency proof between snapshot roots y and z.
+3. Get range of historical values between snapshots y and z.
+
+## Client Verification
+
+Key Transparency clients store:
+1. The root of the log of map roots. This ensures that the client is using the same snapshots as the rest of the world.
+1. The root of every (proven consistent) mini-log they have queried. This ensures that snapshots represent append-only representations of the world.
+
+When querying, Key Transparency clients ask for proof that the current snapshot and value are consistent with previous values that the client has observed.
+
+## Efficiency Innovations
+
+Generating append-only proofs for large sparse Merkle trees is not efficient.
+An append-only proof between two snapshots containing `N` changes per snapshot in a map of size `M` over `T` snapshots contains roughly `O(T * N log M)` hashes.
+1. Instead of verifying that the entire map is an append-only operation from previous values, we isolate the work of verification to individual sub-trees. `O(T * 1 log M)`
+1. Rather than verifying every single snapshot, we only verify the snapshots that the sender and receiver used. `O(1 * 1 log M)`
+1. But the snapshots that the sender and receiver used are unknown, so we use a meet-in-the-middle algorithm to sample log T of them. `O( log T * log M)`
+
+
+# Work Plan
+
+1. Migrate to use Trillian Logs mini logs for user updates
+    1. Switch from `int64` treeIDs to `[]byte` to support billions of users.
```

Comment:
`int64` seems still enough to support billions. Do you think of making `[]byte` the VRF? Is there a way to avoid this, by having each client know their own `int64 treeID` and their friends IDs?

Reply:
The reason is to support server adversarially selected IDs. For 128 bits of security, we need 256 bits of ID.  Maybe we could use the 64 bit VRF prefix for storage but use the full VRF in the map?